airflow/utils/timezone.py

```diff
@@ -195,11 +201,11 @@ def coerce_datetime(v: None) -> None:
 
 
 @overload
-def coerce_datetime(v: dt.datetime) -> DateTime:
+def coerce_datetime(v: dt.datetime) -> dt.datetime:
     ...
 
 
-def coerce_datetime(v: Optional[dt.datetime]) -> Optional[DateTime]:
+def coerce_datetime(v: Optional[dt.datetime]) -> Optional[dt.datetime]:
```

Comment:
This shouldn’t be necessary; `coerce_datetime` always returns a `pendulum.DateTime`. Also I believe this will cause some errors to pop up elsewhere because there are functions expecting this to return a `pendulum.DateTime` specifically.

Reply:
See the error here when changed back
```
airflow/utils/timezone.py:213: error: Incompatible return value type (got "datetime", expected
"Optional[DateTime]")
            return v if v.tzinfo else make_aware(v)
                   ^
Found 1 error in 1 file (checked 60 source files)
```

Reply:
I think we need additional overload of `make_aware` with `DateTime` that should fix it. https://mypy.readthedocs.io/en/stable/more_types.html#function-overloading

Reply:
I tried adding an additional overload of `make_aware` but mypy errors that it'll never be matched because of the arguments.
```
airflow/utils/timezone.py:113: error: Overloaded function signature 3 will
never be matched: signature 2's parameter type(s) are the same or broader
    def make_aware(value: dt.datetime, timezone: Optional[dt.tzinfo] = N...
    ^
```

Reply:
OK. I figured it out. The preoblm is that DateTime derives from dt.datetime. And we have to specify DateTime overloads before datetimes ones. 

What worked for me:

* make_aware overloads:

```
@overload
def make_aware(value: None, timezone: Optional[dt.tzinfo] = None) -> None:
    ...


@overload
def make_aware(value: DateTime, timezone: Optional[dt.tzinfo] = None) -> DateTime:
    ...


@overload
def make_aware(value: dt.datetime, timezone: Optional[dt.tzinfo] = None) -> dt.datetime:
    ...


def make_aware(value: Optional[dt.datetime], timezone: Optional[dt.tzinfo] = None) -> Optional[dt.datetime]:
```

* coerce_datetime overloads:

```
@overload
def coerce_datetime(v: None) -> None:
    ...


@overload
def coerce_datetime(v: DateTime) -> DateTime:
    ...


@overload
def coerce_datetime(v: dt.datetime) -> DateTime:
    ...


def coerce_datetime(v: Optional[dt.datetime]) -> Optional[DateTime]:
```